Initial version of Project Concept for CV32E40PX (#634)
openhwgroup/programs: Project-Descriptions-and-Plans/CV32E4_TBD/Project_Concept_for_CV32E4_TBD_Nov-27_2023.md

> # OpenHW Project Concept Proposal: CV32E4_TBD

Reviewer: I would prefer having a name here and then changing it later so that we can talk about it more agile

Author: ```suggestion
# OpenHW Project Concept Proposal: CV32E40PX
```

> # OpenHW Project Concept Proposal: CV32E4_TBD

Author: ```suggestion
# OpenHW Project Concept Proposal: CV32E40PX
```

*The PC proposal introduces the project and explains the market drivers and the "why"*

# Title of Project - "CV32E4_TBD"
# Project Concept Proposal
## Date of proposal - 2023-11-27
## Author(s) - Manuel PEZZIN (CEA)

## High Level Summary of project, project components, and deliverables

> This proposal is a continuation of CV32E40Pv2 project.

Reviewer: this creates confusion, cause you are actually making a new project as it is not a v3 project -

Main goal is to support official RVB (and a part of RVK) and (upcoming) RVP

> extensions while keeping backward compatibility with CV32E40Pv1/v2 (eg XPULP

Reviewer: are you planning to be LEC with v1? or SEQ? or neither of? what is the TRL you expect to achieve?

Author: We want to keep functional equivalence only. We could try to keep LEC with v1 for basic configuration (without XPULP, RVB, RVP) but I don't want LEC being a strong requirement.

Reviewer: ok makes sense - let's aim for SEQ maybe?

Author: SEQ will definitely be useful during the first round of RTL modifications. This first step will mainly consist in ALU code refactoring (resource sharing, deduplication) and preparing it for addition of new features from RVB/RVP.
Maintaining SEQ after this first phase would require having a design configuration that is strictly equivalent to the most complete one from CV32E40Pv2. I don't know how to manage this properly yet but we should think about it.

> This proposal is a continuation of CV32E40Pv2 project.
> Main goal is to support official RVB (and a part of RVK) and (upcoming) RVP
> extensions while keeping backward compatibility with CV32E40Pv1/v2 (eg XPULP

Author: What about this instead ?
```suggestion
Main goal of this project is to provide a new CPU core that supports official RVB (and a part of RVK) and (upcoming) RVP
extensions while keeping functional backward compatibility with CV32E40Pv1/v2 (eg XPULP
```

custom extensions are kept and even extended).
Additionnal features may also be considered, like:
* CV-X-IF support
* CLIC support

> * Improved OBI interface support (bus error in particular)

Reviewer: .ok so no LEC :) - or at least only if bus error is 0

Author: Same comment as above. I think LEC may be envisioned with a case analysis on bus error input

Reviewer: makes sense to me

This new project will keep functionnal backward compatibility with CV32E40Pv2,
but with relaxed rules compared to previous project. In particular, formal equivalence

> with previous design will not be required to make design work easier.

Reviewer: but how about v1?

Author: As v2 is already backward-compatible with v1, I consider this will be (implicitly) the case, provided v1 configuration (no XPULP, no RVB, no RVP, etc.) remains in verification plan.

Reviewer: ok, thx - I think we can add 1 line to say that explicitly :)

> with previous design will not be required to make design work easier.

Reviewer 2: ```suggestion
with the previous design will not be required to make design work easier.
```

Reviewer 2: What is the cost of not requiring formal equivalence? Should comment on this and not just the motivation.

Due to the strict backward compatibility rules of CV32E40Pv2, this new core is not
named CV32E40Pv3 and a new naming should be found.

## Summary of market or input requirements

### Known market/project requirements at PC gate

> 

Reviewer 2: The list of requirements is good. However, we are missing an explanation of where these requirements arise from. Can you please give an overall summary of the objective of the new core, that is what segment will it fill that differs from E40Pv2, E40X, etc.?

Author: The objective is to have a core tailored for embedded real-time control/DSP workloads.
My initial intent was was to propose a work plan for a new revision of one of these 2 cores that would support RVB and RVP extensions.
After an in-depth review, the outcome was that I would need to mix features from both designs.
Using CV32E40Pv2 as stating point would make this "merge" easier because the SIMD datapath is verified and will be a good starting point for adding RVP support.
Backporting RVB, RVK, CLIC and CV-X-IF from E40X was judged easier than re-introducing XPULP features to E40X

Author: Is it sufficient if I move the explication text that is at the end back to the first section ?

Reviewer 2: @MPEZZIN that is fine

> **Madatory features and design changes**

Reviewer 2: ```suggestion
**Mandatory features and design changes**
```

Reviewer 2: What do you mean by mandatory in this context?

Introduce a dual opcode scheme for XPULP ALU instructions that have a strict RVB/RVP equivalent:
* Multiply-accumulate
* General ALU extensions
* Bit manipulation
* SIMD (16 and 8-bit data)
note : this dual opcode scheme allows backard compatibility and open the door to smooth transition
of existing programs from XPULP to RVB/RVP, if needed/relevant.

Add support for RISC-V official bit manipulation instructions:
* RVB zba (address manipulation) will be supported
* RVB zbb (basic bit manipulations) will be supported
* RVB zbs (single bit instructions) will be supported
* RVB zbc (carry-less multiplications) : may be supported, to be confirmed if relevant for DSP applications
* RVK zbkb (Bitmanip instructions for Cryptography) : pack, packh, packw, brev8, zip, unzip additionnal instructions may be supported, if relevant for DSP applications.
* RVK zbkx (Crossbar permutation instructions) : xpem4 and xperm8 additionnal instructions may be supported, if relevant for DSP applications.
* RVK zkn, zks, zkr and zkt are specific to cryptographic computations and will not be supported

**Optionnal design changes**

Introduce XPULPv3 custom extentions (preliminary, feature list may evolve):

> * ALU instructions that have a strict RVB/RVP equivalent opcode will be deprecated, new XPULPv3 opcodes will be RVB/RVP ones instead

Reviewer: Here it is not clear - previously we wrote that we keep binary compatibility, here we say that we deprecate some of those - is it something you want to do in the future, with a parameter? and if so, how do you verify the CPU with xpulp and xpulpv3 extensions? do you freeze both versions in 2 different stages?

Author: Advertising a feature as deprecated doesn't mean it is unmaintained or removed in the current version.
This is more to inform for the end user that this feature will be removed in a future version and that SW should be updated to use new equivalent features if a long term product plan (using future version of this core) is envisioned.

Author: Regarding XPULPv3, maintaining dual instruction encoding for the first version will allow reusing current XPULP verification for the first version of the core. Then, the new instruction encoding and new instructions will be managed as an add-on to existing test suite.
Deprecated encodings will be removed from test suite of a future revision of the core, when actual removal is decided.

Reviewer: but will the XPULPv3 be part of this core or not? and if so, with a parameter? what do you plan to verify as first?

Author: Yes XPULPv3 will be part of this core. What I have in mind is to extend the meaning of COREV_PULP parameter to play with features granularity. For example this could be:
- 0 : no XPULP feature at all
- 1 : full XPULPv2 support
- 2 : XPULPv3 support only
- 3 : XPULPv2 + XPULPv3 support
Another option could be a split of XPULP extensions into subset that have a corresponding gcc flag (HW loop, load/stores, SIMD, bitmanip, complex...) and add new flags for v3 support.

In any case, what I plan to verify first is the full XPULPv2 support.

* Deprecate single operand *abs* instruction in favor of dual operand *pdif* (RVP superset of abs useful for manhattan distance computation)
* Deprecate single operand *cv.cplxconj* instruction in favor of dual operand *pas* (RVP parallel add-sub)
* Deprecate *cv.cplxmul.r* and *cv.cplxmul.i* in favor of a single instruction *cv.cplxmul* that computes both parts of complex multiplication
* Introduce hermitian form of *cv.cplxmul* (rs1 x conj(rs2)) ? (widely used in DSP computation for telecom applications)
* *list to be completed / updated*

Add CV-X-IF interface:
* End user product may require additionnal appliance-specific custom instructions, these instructions should be implemented in a co-processor.

> 

Reviewer 2: What kind of end user product  may have requirements for application-specific custom instructions arising, keep in mind this is the project CONCEPT so we are capturing "Why do we need to do this project?".. "What is its application").

On CV-X-IF to summarize, are saying that the need for X-IF will be determined at Project Launch stage? Can you give some insight into how this will be decided?

Author: The kind of systems I have in mind (and that we are working on already with my colleagues) are not general purpose microcontrollers. The embedded processor is a deeply embedded building bloc of a larger System on Chip that is also composed of application-specific HW IPs. These IPs are mainly peripherals connected to system bus but we are also considering using custom instructions when a tight integration with program flow is relevant. CV-X-IF is a perfect candidate to do that since we don't need to modify CPU core.

Floating Point Unit:

> * keep current CV-FPU integration or move it to CV-X-IF?

Reviewer: I would keep it as it is verified

Author: I'm in line with you. But I don't plan to put any design effort on the FPU. We will keep CV32E40Pv2 RTL for this part of the design.
Is it OK for you ?

Author: ```suggestion
* keep current CV-FPU integration unchanged as much as possible. If a design modification is envisioned, sequential equivalence with CV32E40Pv2 is required.
```

Add CLIC support:

> * Core-local interrupt controller fits better with embedded real-time requirements than PLIC or any other (non RISC-V compliant)

Reviewer: do you plan to remove CLINT+PLIC for CLIC or have a parameter?

Reviewer 2: Sorry to be repetitive, but please describe a bit more fully why CLIC is needed

Author: I want this core being tailored for hard real time embedded DSP applications. Using PLIC for interrupt management will introduce an additional source of latency or variability that we cannot afford in many workloads. CLIC specification address that requirement (RISC-V).
Motivation could be a copy/paste from chapter 2 of clic specification. I will add a short sentence with a link to this chapter.

Author: And my plan is to backport CLIC support from E40X design

  external interrupt controller that requires a 2-step interrupt handling code.

Design Portability:
* Replace global clock gate by equivalent global enable signal
  * this will make the design more "FPGA-friendly", or compliant with ASIC design rules that forbid any logic addition on clock tree
  * clock gate cell model and/or simulator behaviour may introduce delta cycles on clock tree, leading to RTL/GL simulation mismatch
* EDA tools parser bug work-arounds: current design coding style is standard-compliant, but EDA tools may
  have bugs that prevent proper reading. Design should be parsed with several EDA tools and offending RTL code should be fixed.

> ### Potential future design enhancements

Reviewer 2: We should probably rename this heading in the template as it isn't quite clear. Anyways, are these features thinks you may decide to include within the overall scope of the Project Launch, or are they things that might be added in a future release of this project?

Author: These are things that were discussed with several partners. I don't want them being part of the first release of the core for various reasons (reuse of verification code, backward SW compatibility, CV-X-IF specification maturity...) but that should be considered for the next release.

* remove deprecated features
* Add Zce extension for further code-size reduction (not required for Multicore cluster).
* Add bus attributes support.
* move bitmanip, SIMD and complex number subsets of XPULP, RVB and RVP to a CV-X-IF compliant co-processor IP
  * pros : this will make these features available to other processors, and would simplify verification work.
  * cons : a performance loss is expected on instructions that use register pairs.

## Who would make use of OpenHW output

> Companies needing more performances, less energy consumption or smaller code size for real-time control and/or DSP applications.

Reviewer 2: More in comparison to which cores?

Author: ARM Cortex M family and the likes

## Initial Estimate of Timeline
At Project Concept:

> * Preliminary RTL release expected in 2024 Q1

Reviewer 2: That means RTL will be finished in 3-4 months, is that correct? Let's not call that release as I would anticipate we need to more than complete RTL prior to release

Author: No ! I'm expected RTL design to start in 2024 Q1. This is the expected date for our first push.

```suggestion
* Preliminary RTL code drop expected in 2024 Q1
```

* RTL feature-complete expected in end of 2024 Q4
* RTL freeze and verification completed expected in 2025 Q4

> ## Explanation of why OpenHW should do this project

Reviewer 2: Suggest we move most of the content here up to the front of the document as it kind of motivates what follows. The template section is actually intended to be "why should OpenHW specifically do this project", again the template needs a tweak. In the case of this project, it clearly is within the scope of OpenHW's work so we don't need to justify that too much. If part of this work is being done by Tristan then that could be explained here

Author: I confirm that I misunderstood the purpose of this section.
In fact, this text was in the first section in my original draft. Then I moved it here and made a summary of it in the first section instead. No problem to move it back to the first section.

Author: This work is a contribution from TRISTAN project but we (CEA) clearly stated to the consortium that we will do it as OpenHW Group Member.
I also expect this work to continue after TRISTAN is finished.
Maybe we should add the official disclaimer from TRISTAN consortium here ?

Main goal is to support official RVB (and a part of RVK) and (upcoming) RVP
extensions while keeping backward compatibility with CV32E40Pv1/v2.
One key advantage of supporting official RVB/RVP extensions is versatility and
compatibility with a larger set of framework and applications. This will also allow
users to use upstream version of compilation toolchain.

XPULP custom extensions may also be extended to benefit from design enhancements
introduced for RVB and RVP support.

Additionnal features may also be considered, like CV-X-IF and CLIC support.

Improved OBI interface support (bus error in particular) is also envisionned.

Keeping backward compatibility will allow existing users of CV32E40Pv2 to migrate
smoothly their existing products to this new version of the core, while keeping
the benefits of existing custom extensions.
In addition, keeping XPULP custom extensions will offer improvements to address
manipulation, flow control and complex numbers computations.

## Industry landscape: description of competing, alternative, or related efforts in the industry
ARM Cortex M4

## OpenHW Members/Participants committed to participate

> * CEA - RVB/RVP/XPLULP Design & Verification

Reviewer 2: What can we do to bring other OpenHW members into participation? We don't normally want a project with just one member driving. That can be a PL deliverable item.

Author: Davide and his colleagues from EPFL intend to work on CV-X-IF. Work has started already here : https://github.com/esl-epfl/cv32e40px
I don't know yet if Dolphin or other partners would be interested to participate.

## Project Leader(s)
### Technical Project Leader(s)
At of Project Concept, led by
* Manuel Pezzin, CEA

### Project Manager, if a PM is designated
None designated

## Next steps/Investigation towards Project Launch (**PC only**)

Verification asumptions (to be validated):

> * functionnal backward compatibility with CV32E40Pv2 will allow reusing existing test suite

Reviewer 3: Maybe not totally true any more.
I close this remark.

real-time control/DSP workloads.

Target TRL is TRL4 or TRL5 (idealy TRL5 if we are able to find enough verification resources).

Main goal of this project is to provide a new CPU core that supports official RVB
(and a part of RVK) and (upcoming) RVP extensions while keeping functional backward
compatibility with CV32E40Pv1/v2 (eg XPULP custom extensions are kept and even extended).

Additionnal features may also be considered, like:
* CV-X-IF support
* CLIC support
* Improved OBI interface support (bus error in particular)

Thus, it will extend existing features from CV32E40P with a subset of CV32E40X ones,
and will extend them further with RVP support in particular.

This new project will keep functional backward compatibility with CV32E40P (v1/v2),
but with relaxed rules on changes compared to the previous project.
In particular, formal logic equivalence with the previous design will not be required
to make design work easier. We will try as much as possible to maintain a formal
sequential equivalence for backward-compatible configurations.

Due to the strict backward compatibility rules of CV32E40Pv2, this new core is not

> named CV32E40Pv3 and a new naming should be found.

Author: This last sentence was removed. Text above is clear enough.
